op-batcher/batcher/channel_manager_test.go

```diff
@@ -483,3 +484,174 @@ func TestChannelManager_ChannelCreation(t *testing.T) {
+	chooseBlobs bool
+}
+
+func (f *FakeDynamicEthChannelConfig) ChannelConfig() ChannelConfig {
+	if f.chooseBlobs {
+		return f.blobConfig
+	}
+	return f.calldataConfig
+}
+
+func newFakeDynamicEthChannelConfig(lgr log.Logger,
+	reqTimeout time.Duration) *FakeDynamicEthChannelConfig {
+
+	calldataCfg := ChannelConfig{
+		MaxFrameSize:    120_000 - 1,
+		TargetNumFrames: 1,
+	}
+	blobCfg := ChannelConfig{
+		MaxFrameSize:    eth.MaxBlobDataSize - 1,
+		TargetNumFrames: 3, // gets closest to amortized fixed tx costs
+		UseBlobs:        true,
+	}
+	calldataCfg.InitRatioCompressor(1, derive.Brotli)
+	blobCfg.InitRatioCompressor(1, derive.Brotli)
+
+	return &FakeDynamicEthChannelConfig{
+		chooseBlobs: false,
+		DynamicEthChannelConfig: *NewDynamicEthChannelConfig(
+			lgr,
+			reqTimeout,
+			&mockGasPricer{},
+			blobCfg,
+			calldataCfg),
+	}
+}
+
+// TestChannelManager_TxData seeds the channel manager with blocks and triggers the
+// blocks->channels pipeline multiple times. Values are chosen such that a channel
+// is created under one set of market conditions, and then submitted under a different
+// set of market conditions. The test asserts that the DA type is changed at channel
+// submission time.
+func TestChannelManager_TxData(t *testing.T) {
+
+	type TestCase struct {
+		name                            string
+		chooseBlobsWhenChannelCreated   bool
+		chooseBlobsWhenChannelSubmitted bool
+	}
+
+	tt := []TestCase{
+		{"blobs->blobs", true, true},
+		{"calldata->calldata", false, false},
+		{"blobs->calldata", true, false},
+		{"calldata->blobs", false, true},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			l := testlog.Logger(t, log.LevelCrit)
+
+			cfg := newFakeDynamicEthChannelConfig(l, 1000)
+
+			cfg.chooseBlobs = tc.chooseBlobsWhenChannelCreated
+			m := NewChannelManager(l, metrics.NoopMetrics, cfg, defaultTestRollupConfig)
+			require.Equal(t, tc.chooseBlobsWhenChannelCreated, m.defaultCfg.UseBlobs)
+
+			// Seed channel manager with a block
+			rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+			blockA := derivetest.RandomL2BlockWithChainId(rng, 10, defaultTestRollupConfig.L2ChainID)
+			m.blocks = []*types.Block{blockA}
+
+			// Call TxData a first time to trigger blocks->channels pipeline
+			_, err := m.TxData(eth.BlockID{})
+			require.ErrorIs(t, err, ErrInsufficientData)
+
+			// The test requires us to have something in the channel queue
+			// at this point, but not yet ready to send and not full
+			require.NotEmpty(t, m.channelQueue)
+			require.False(t, m.channelQueue[0].IsFull())
+
+			// Simulate updated market conditions
+			// by possibly flipping the state of the
+			// fake channel provider
+			l.Info("updating market conditions", "chooseBlobs", tc.chooseBlobsWhenChannelSubmitted)
+			cfg.chooseBlobs = tc.chooseBlobsWhenChannelSubmitted
+
+			// Add a block and call TxData until
+			// we get some data to submit
+			var data txData
+			for {
+				m.blocks = []*types.Block{blockA}
```

Comment:
You're using Brotli compression and always add the same block, so this compresses probably very well and can take a lot of iterations to fill up. Maybe better generate a new random block every iteration to fill up the channel more quickly?

Reply:
Good point. I the end I went with the NoneCompressor [2a9b7ef](https://github.com/ethereum-optimism/optimism/pull/12002/commits/2a9b7efd1d58d3d70a255c889e8b3552df59239f) and also increased the size of the (same) block to fill the channel more quickly (after 3 blocks) [7d63970](https://github.com/ethereum-optimism/optimism/pull/12002/commits/7d639707e7632deebf8d4f37bd05366bae1567bb)

Reply:
Even better! The NoneCompressor is a good choice in tests.